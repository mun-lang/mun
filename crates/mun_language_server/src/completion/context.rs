use mun_hir::{
    semantics::{Semantics, SemanticsScope},
    AstDatabase,
};
use mun_syntax::{ast, utils::find_node_at_offset, AstNode, SyntaxNode, TextRange, TextSize};
use ra_ap_text_edit::Indel;

use crate::{db::AnalysisDatabase, FilePosition};

/// A `CompletionContext` is created to figure out where exactly the cursor is.
pub(super) struct CompletionContext<'a> {
    pub sema: Semantics<'a>,
    pub scope: SemanticsScope<'a>,
    pub db: &'a AnalysisDatabase,

    // TODO: Add this when it is used
    //pub position: FilePosition,
    /// True if the context is currently at a trivial path.
    pub is_trivial_path: bool,

    /// True if the context is currently on a parameter
    pub is_param: bool,

    /// True if we're at an `ast::PathType`
    pub is_path_type: bool,

    /// The receiver if this is a field or method access, i.e. writing
    /// something.$0
    pub dot_receiver: Option<ast::Expr>,
}

impl<'a> CompletionContext<'a> {
    /// Tries to construct a new `CompletionContext` with the given database and
    /// file position.
    pub fn new(db: &'a AnalysisDatabase, position: FilePosition) -> Option<Self> {
        let sema = Semantics::new(db);

        let original_file = sema.parse(position.file_id);

        // Insert a fake identifier to get a valid parse tree. This tree will be used to
        // determine context. The actual original_file will be used for
        // completion.
        let file_with_fake_ident = {
            let parse = db.parse(position.file_id);
            let edit = Indel::insert(position.offset, String::from("intellijRulezz"));
            parse.reparse(&edit).tree()
        };

        // Get the current token
        let token = original_file
            .syntax()
            .token_at_offset(position.offset)
            .left_biased()?;

        let scope = sema.scope_at_offset(&token.parent()?, position.offset);

        let mut context = Self {
            sema,
            scope,
            db,
            // TODO: add this when it is used
            //position,
            is_trivial_path: false,
            is_param: false,
            is_path_type: false,
            dot_receiver: None,
        };

        context.fill(
            &original_file.syntax().clone(),
            file_with_fake_ident.syntax().clone(),
            position.offset,
        );

        Some(context)
    }

    /// Examine the AST and determine what the context is at the given offset.
    fn fill(
        &mut self,
        original_file: &SyntaxNode,
        file_with_fake_ident: SyntaxNode,
        offset: TextSize,
    ) {
        // First, let's try to complete a reference to some declaration.
        if let Some(name_ref) = find_node_at_offset::<ast::NameRef>(&file_with_fake_ident, offset) {
            if is_node::<ast::Param>(name_ref.syntax()) {
                self.is_param = true;
                return;
            }

            self.classify_name_ref(original_file, name_ref);
        }
    }

    /// Classifies an `ast::NameRef`
    fn classify_name_ref(&mut self, original_file: &SyntaxNode, name_ref: ast::NameRef) {
        let parent = match name_ref.syntax().parent() {
            Some(it) => it,
            None => return,
        };

        // Complete references to declarations
        if let Some(segment) = ast::PathSegment::cast(parent.clone()) {
            let path = segment.parent_path();

            self.is_path_type = path
                .syntax()
                .parent()
                .and_then(ast::PathType::cast)
                .is_some();

            if let Some(segment) = path.segment() {
                if segment.has_colon_colon() {
                    return;
                }
            }

            self.is_trivial_path = true;
        }

        // Complete field expressions
        if let Some(field_expr) = ast::FieldExpr::cast(parent) {
            // The receiver comes before the point of insertion of the fake
            // ident, so it should have the same range in the non-modified file
            self.dot_receiver = field_expr
                .expr()
                .map(|e| e.syntax().text_range())
                .and_then(|r| find_node_with_range(original_file, r));
        }
    }
}

/// Returns true if the given `node` or one if its parents is of the specified
/// type.
fn is_node<N: AstNode>(node: &SyntaxNode) -> bool {
    match node.ancestors().find_map(N::cast) {
        None => false,
        Some(n) => n.syntax().text_range() == node.text_range(),
    }
}

/// Returns a node that covers the specified range.
fn find_node_with_range<N: AstNode>(syntax: &SyntaxNode, range: TextRange) -> Option<N> {
    syntax.covering_element(range).ancestors().find_map(N::cast)
}
