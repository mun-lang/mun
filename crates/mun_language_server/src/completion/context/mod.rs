#![allow(dead_code)]

mod analysis;

use mun_hir::{
    semantics::{PathResolution, Semantics, SemanticsScope},
    AstDatabase, Ty,
};
use mun_syntax::{ast, AstNode, SyntaxNode};
use ra_ap_text_edit::Indel;

use crate::{
    completion::context::analysis::{analyze, AnalysisResult},
    db::AnalysisDatabase,
    FilePosition,
};

/// A `CompletionContext` is created to figure out where exactly the cursor is.
pub(super) struct CompletionContext<'a> {
    pub sema: Semantics<'a>,
    pub scope: SemanticsScope<'a>,
    pub db: &'a AnalysisDatabase,
    // TODO: Add this when it is used
    //pub position: FilePosition,
}

/// Information about the identifier that we are currently completing.
#[derive(Debug)]
pub(super) enum CompletionAnalysis {
    NameRef(NameRefContext),
}

/// The identifier to complete is a name reference.
#[derive(Debug)]
pub(super) struct NameRefContext {
    /// `NameRef` syntax in the original file
    pub(super) name_ref: Option<ast::NameRef>,
    pub(super) kind: NameRefKind,
}

/// The kind of the `NameRef` we are completing.
#[derive(Debug)]
pub(super) enum NameRefKind {
    Path(PathCompletionContext),
    DotAccess(DotAccess),
}

/// Information about the field or method access we are completing.
#[derive(Debug)]
pub(crate) struct DotAccess {
    pub(crate) receiver: Option<ast::Expr>,
    pub(crate) receiver_ty: Option<Ty>,
}

/// The state of the path we are currently completing.
#[derive(Debug)]
pub(super) struct PathCompletionContext {
    /// The type of path we are completing.
    pub(super) kind: PathKind,

    /// The qualifier of the current path.
    pub(super) qualified: Qualified,

    /// Whether the qualifier comes from a use tree parent or not
    pub(super) use_tree_parent: bool,
}

#[derive(Debug)]
pub(super) enum Qualified {
    /// No path qualifier, this is a bare path, e.g. `foo`
    No,

    /// The path has a qualifier, e.g. `foo` in `foo::bar`
    With {
        /// The path that is being completed
        path: ast::Path,

        /// The resolution of the path that is being completed
        resolution: Option<PathResolution>,
    },

    /// The path has an absolute qualifier, e.g. `::foo`
    Absolute,
}

/// The kind of path we are completing right now.
#[derive(Debug)]
pub(super) enum PathKind {
    Expr(PathExprContext),
    Use,
    SourceFile,
}

#[derive(Debug)]
pub(super) struct PathExprContext {}

impl<'a> CompletionContext<'a> {
    /// Tries to construct a new `CompletionContext` with the given database and
    /// file position.
    pub fn new(
        db: &'a AnalysisDatabase,
        position: FilePosition,
    ) -> Option<(Self, CompletionAnalysis)> {
        let sema = Semantics::new(db);

        let original_file = sema.parse(position.file_id);
        // Insert a fake identifier to get a valid parse tree. This tree will be used to
        // determine context. The actual original_file will be used for
        // completion.
        let file_with_fake_ident = {
            let parse = db.parse(position.file_id);
            let edit = Indel::insert(position.offset, String::from("intellijRulezz"));
            parse.reparse(&edit).tree()
        };

        // Get the current token
        let original_token = original_file
            .syntax()
            .token_at_offset(position.offset)
            .left_biased()?;

        // Analyze the context of the completion request
        let AnalysisResult { analysis } = analyze(
            &sema,
            original_file.syntax().clone(),
            file_with_fake_ident.syntax().clone(),
            position.offset,
        )?;

        let scope = sema.scope_at_offset(&original_token.parent()?, position.offset);

        let context = Self {
            sema,
            scope,
            db,
            // TODO: add this when it is used
            //position,
        };

        Some((context, analysis))
    }
}

/// Attempts to find `node` inside `syntax` via `node`'s text range.
/// If the fake identifier has been inserted after this node or inside of this
/// node use the `_compensated` version instead.
fn find_opt_node_in_file<N: AstNode>(syntax: &SyntaxNode, node: Option<N>) -> Option<N> {
    find_node_in_file(syntax, &node?)
}

/// Attempts to find `node` inside `syntax` via `node`'s text range.
/// If the fake identifier has been inserted after this node or inside of this
/// node use the `_compensated` version instead.
fn find_node_in_file<N: AstNode>(syntax: &SyntaxNode, node: &N) -> Option<N> {
    let syntax_range = syntax.text_range();
    let range = node.syntax().text_range();
    let intersection = range.intersect(syntax_range)?;
    syntax
        .covering_element(intersection)
        .ancestors()
        .find_map(N::cast)
}
