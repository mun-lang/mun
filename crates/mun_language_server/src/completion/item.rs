use crate::SymbolKind;

/// A `CompletionItem` describes a single completion variant in an editor.
#[derive(Clone, Debug)]
pub struct CompletionItem {
    /// Used for tests to filter certain type of completions
    #[allow(unused)]
    pub completion_kind: CompletionKind,

    /// Label in the completion pop up which identifies completion.
    pub label: String,

    /// The type of completion
    pub kind: Option<CompletionItemKind>,

    /// Additional info to show in the UI pop up.
    pub detail: Option<String>,
}

#[derive(Debug, PartialEq, Eq, Copy, Clone)]
pub enum CompletionKind {
    /// Your usual "complete all valid identifiers".
    Reference,
    BuiltinType,
}

/// Type of completion used to provide hints to the user.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[allow(unused)]
pub enum CompletionItemKind {
    SymbolKind(SymbolKind),
    Attribute,
    Binding,
    BuiltinType,
    Keyword,
    Method,
    Snippet,
    UnresolvedReference,
}

impl CompletionItemKind {
    /// Returns a tag that describes the type of item that was completed. This
    /// is only used in tests, to be able to distinguish between items with
    /// the same name.
    #[cfg(test)]
    pub(crate) fn tag(&self) -> &'static str {
        match self {
            CompletionItemKind::SymbolKind(kind) => match kind {
                SymbolKind::Field => "fd",
                SymbolKind::Function => "fn",
                SymbolKind::Local => "lc",
                SymbolKind::Module => "md",
                SymbolKind::Struct => "st",
                SymbolKind::TypeAlias => "ta",
            },
            CompletionItemKind::Attribute => "at",
            CompletionItemKind::Binding => "bn",
            CompletionItemKind::BuiltinType => "bt",
            CompletionItemKind::Keyword => "kw",
            CompletionItemKind::Method => "me",
            CompletionItemKind::Snippet => "sn",
            CompletionItemKind::UnresolvedReference => "??",
        }
    }
}

impl From<SymbolKind> for CompletionItemKind {
    fn from(kind: SymbolKind) -> Self {
        CompletionItemKind::SymbolKind(kind)
    }
}

impl CompletionItem {
    /// Constructs a [`Builder`] to build a `CompletionItem` with
    pub fn builder(kind: CompletionKind, label: impl Into<String>) -> Builder {
        Builder {
            label: label.into(),
            kind: None,
            completion_kind: kind,
            detail: None,
        }
    }
}

/// A builder for a `CompletionItem`. Constructed by calling
/// [`CompletionItem::builder`].
pub struct Builder {
    label: String,
    completion_kind: CompletionKind,
    kind: Option<CompletionItemKind>,
    detail: Option<String>,
}

impl Builder {
    /// Completes building the `CompletionItem` and returns it
    pub fn finish(self) -> CompletionItem {
        CompletionItem {
            completion_kind: self.completion_kind,
            label: self.label,
            kind: self.kind,
            detail: self.detail,
        }
    }

    /// Sets the type of the completion
    pub fn kind(mut self, kind: impl Into<CompletionItemKind>) -> Builder {
        self.kind = Some(kind.into());
        self
    }

    /// Set the details of the completion item
    pub fn detail(mut self, detail: impl Into<String>) -> Builder {
        self.detail = Some(detail.into());
        self
    }
}
