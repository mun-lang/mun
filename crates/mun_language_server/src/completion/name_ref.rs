use super::{dot, expr, CompletionContext, Completions, NameRefContext, NameRefKind, PathKind};

/// Generate completions for a name reference.
#[allow(clippy::single_match)]
pub(super) fn complete_name_ref(
    completions: &mut Completions,
    ctx: &CompletionContext<'_>,
    NameRefContext { kind, .. }: &NameRefContext,
) {
    match kind {
        NameRefKind::Path(path_ctx) => match &path_ctx.kind {
            PathKind::Expr(expr_ctx) => {
                expr::complete_expr_path(completions, ctx, path_ctx, expr_ctx);
            }
            _ => {}
        },
        NameRefKind::DotAccess(dot_access) => {
            dot::complete_dot(completions, ctx, dot_access);
        }
    }
}
