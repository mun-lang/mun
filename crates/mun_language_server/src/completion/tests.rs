#[cfg(test)]
mod test {
    use crate::completion::test_utils::completion_relevance_string;

    #[test]
    fn test_locals_first() {
        insta::assert_snapshot!(completion_relevance_string(
            r#"
            fn a() {};

        fn foo(bar: u32) {
            let zinq = 0;
            z$0
        }
        "#
        ), @r###"
        lc zinq i32
        lc bar  u32
        fn foo  -> ()
        fn a    -> ()
        "###);
    }
}
