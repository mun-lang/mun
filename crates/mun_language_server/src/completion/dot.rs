use mun_db::Upcast;

use super::{CompletionContext, Completions};

/// Complete dot accesses, i.e. fields. Adds `CompletionItems` to `result`.
pub(super) fn complete_dot(result: &mut Completions, ctx: &CompletionContext<'_>) {
    // Get the expression that we want to get the fields of
    let dot_receiver = match &ctx.dot_receiver {
        Some(expr) => expr,
        _ => return,
    };

    // Figure out the type of the expression
    let receiver_ty = match ctx.sema.type_of_expr(dot_receiver) {
        Some(ty) => ty,
        _ => return,
    };

    // Get all the fields of the expression
    if let Some(strukt) = receiver_ty.as_struct() {
        for field in strukt.fields(ctx.db.upcast()) {
            result.add_field(ctx, field);
        }
    };
}

#[cfg(test)]
mod tests {
    use crate::completion::{test_utils::completion_string, CompletionKind};

    #[test]
    fn test_struct_fields() {
        insta::assert_snapshot!(completion_string(
            r#"
        struct FooBar {
            foo: i32,
            bar: i32,
        };
        
        fn foo() {
            let bar = FooBar { foo: 0, bar: 0 };
            bar.$0
        }
        "#,
            Some(CompletionKind::Reference)
        ));
    }

    #[test]
    fn test_tuple_struct() {
        insta::assert_snapshot!(completion_string(
            r#"
        struct FooBar(i32, i32)
        
        fn foo() {
            let bar = FooBar(0,0);
            bar.$0
        }
        "#,
            Some(CompletionKind::Reference)
        ));
    }

    #[test]
    fn test_nested_struct() {
        insta::assert_snapshot!(completion_string(
            r#"
        struct Foo { baz: i32 }
        struct Bar(Foo)
        
        fn foo() {
            let bar = Bar(Foo{baz:0});
            bar.0.$0
        }
        "#,
            Some(CompletionKind::Reference)
        ));
    }

    #[test]
    fn test_incomplete_struct() {
        insta::assert_snapshot!(completion_string(
            r#"
        struct Foo { bar: i32 }
        
        fn foo() {
            let bar = Foo;
            bar.$0
        }
        "#,
            Some(CompletionKind::Reference)
        ));
    }
}
