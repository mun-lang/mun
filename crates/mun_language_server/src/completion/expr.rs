use mun_hir::{
    method_resolution::{AssociationMode, MethodResolutionCtx},
    semantics::PathResolution,
    AssocItemId, ModuleDef,
};

use super::{CompletionContext, Completions, PathCompletionContext, PathExprContext, Qualified};

pub(super) fn complete_expr_path(
    result: &mut Completions,
    ctx: &CompletionContext<'_>,
    PathCompletionContext { qualified, .. }: &PathCompletionContext,
    _expr_ctx: &PathExprContext,
) {
    match qualified {
        Qualified::With {
            resolution: Some(PathResolution::Def(ModuleDef::Struct(s))),
            ..
        } => {
            let ty = s.ty(ctx.db);
            MethodResolutionCtx::new(ctx.db, ty.clone())
                .with_association(AssociationMode::WithoutSelf)
                .collect(|item, _visible| {
                    match item {
                        AssocItemId::FunctionId(f) => result.add_function(ctx, f.into(), None),
                    };
                    None::<()>
                });
        }
        Qualified::No => {
            // Iterate over all items in the current scope and add completions for them
            ctx.scope.visit_all_names(&mut |name, def| {
                result.add_resolution(ctx, name.to_string(), &def);
            });
        }
        _ => {}
    }
}

#[cfg(test)]
mod tests {
    use crate::completion::{test_utils::completion_string, CompletionKind};

    #[test]
    fn test_local_scope() {
        insta::assert_snapshot!(completion_string(
            r#"
        fn foo() {
            let bar = 0;
            let foo_bar = 0;
            f$0
        }
        "#,
            Some(CompletionKind::Reference)
        ));
    }

    #[test]
    fn test_associate_function() {
        insta::assert_snapshot!(completion_string(
            r#"
        struct Foo;

        impl Foo {
            fn new() -> Self {
                Self
            }
        }

        fn foo() {
            let bar = Foo::$0;
        }
        "#,
            Some(CompletionKind::Reference)
        ));
    }
}
