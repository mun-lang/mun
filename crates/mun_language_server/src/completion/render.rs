mod function;

use super::{CompletionContext, CompletionItem, CompletionItemKind};
use crate::completion::CompletionKind;
use crate::{db::AnalysisDatabase, SymbolKind};
use function::FunctionRender;
use mun_hir::{semantics::ScopeDef, HirDisplay};

pub(super) fn render_field(ctx: RenderContext, field: mun_hir::Field) -> CompletionItem {
    Render::new(ctx).render_field(field)
}

pub(super) fn render_fn(
    ctx: RenderContext,
    local_name: Option<String>,
    func: mun_hir::Function,
) -> Option<CompletionItem> {
    Some(FunctionRender::new(ctx, local_name, func)?.render())
}

pub(super) fn render_resolution(
    ctx: RenderContext,
    local_name: String,
    resolution: &ScopeDef,
) -> Option<CompletionItem> {
    Render::new(ctx).render_resolution(local_name, resolution)
}

/// Interface for data and methods required for items rendering.
pub(super) struct RenderContext<'a> {
    completion: &'a CompletionContext<'a>,
}

impl<'a> RenderContext<'a> {
    pub(super) fn new(completion: &'a CompletionContext<'a>) -> RenderContext<'a> {
        RenderContext { completion }
    }

    pub(super) fn db(&self) -> &'a AnalysisDatabase {
        self.completion.db
    }
}

/// Generic renderer for completion items.
struct Render<'a> {
    ctx: RenderContext<'a>,
}

impl<'a> Render<'a> {
    fn new(ctx: RenderContext<'a>) -> Render<'a> {
        Render { ctx }
    }

    /// Constructs a `CompletionItem` for a resolved name.
    fn render_resolution(
        self,
        local_name: String,
        resolution: &ScopeDef,
    ) -> Option<CompletionItem> {
        use mun_hir::ModuleDef::*;

        let completion_kind = match resolution {
            ScopeDef::ModuleDef(PrimitiveType(..)) => CompletionKind::BuiltinType,
            _ => CompletionKind::Reference,
        };

        let kind = match resolution {
            ScopeDef::ModuleDef(Module(_)) => CompletionItemKind::SymbolKind(SymbolKind::Module),
            ScopeDef::ModuleDef(Function(func)) => {
                return render_fn(self.ctx, Some(local_name), *func)
            }
            ScopeDef::ModuleDef(PrimitiveType(_)) => CompletionItemKind::BuiltinType,
            ScopeDef::ModuleDef(Struct(_)) => CompletionItemKind::SymbolKind(SymbolKind::Struct),
            ScopeDef::ModuleDef(TypeAlias(_)) => {
                CompletionItemKind::SymbolKind(SymbolKind::TypeAlias)
            }
            ScopeDef::Local(_) => CompletionItemKind::SymbolKind(SymbolKind::Local),
            ScopeDef::Unknown => {
                let item = CompletionItem::builder(CompletionKind::Reference, local_name)
                    .kind(CompletionItemKind::UnresolvedReference)
                    .finish();
                return Some(item);
            }
        };

        let mut item = CompletionItem::builder(completion_kind, local_name);

        // Add the type for locals
        if let ScopeDef::Local(local) = resolution {
            let ty = local.ty(self.ctx.db());
            if !ty.is_unknown() {
                item = item.detail(ty.display(self.ctx.db()).to_string());
            }
        };

        Some(item.kind(kind).finish())
    }

    /// Constructs a `CompletionItem` for a field.
    fn render_field(&mut self, field: mun_hir::Field) -> CompletionItem {
        let name = field.name(self.ctx.db());
        CompletionItem::builder(CompletionKind::Reference, name.to_string())
            .kind(SymbolKind::Field)
            .detail(field.ty(self.ctx.db()).display(self.ctx.db()).to_string())
            .finish()
    }
}
