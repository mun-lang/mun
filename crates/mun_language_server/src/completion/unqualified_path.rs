use super::{CompletionContext, Completions};

/// Adds completions to `result` for unqualified path. Unqualified paths are
/// simple names which do not refer to anything outside of the current scope:
/// local function names, variables, etc. E.g.:
/// ```mun
/// fn foo() {
///    let foo_bar = 3;
///    foo_$0
/// }
/// ```
pub(super) fn complete_unqualified_path(result: &mut Completions, ctx: &CompletionContext<'_>) {
    // Only complete trivial paths (e.g. foo, not ::foo)
    if !ctx.is_trivial_path {
        return;
    }

    // Iterate over all items in the current scope and add completions for them
    ctx.scope.visit_all_names(&mut |name, def| {
        result.add_resolution(ctx, name.to_string(), &def);
    });
}

#[cfg(test)]
mod tests {
    use crate::completion::{test_utils::completion_string, CompletionKind};

    #[test]
    fn test_local_scope() {
        insta::assert_snapshot!(completion_string(
            r#"
        fn foo() {
            let bar = 0;
            let foo_bar = 0;
            f$0
        }
        "#,
            Some(CompletionKind::Reference)
        ));
    }
}
