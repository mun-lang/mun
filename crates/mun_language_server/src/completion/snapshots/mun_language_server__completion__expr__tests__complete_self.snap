---
source: crates/mun_language_server/src/completion/expr.rs
expression: "completion_string(r#\"\n            struct Foo;\n\n        impl Foo {\n            fn foo(self) {\n                $0\n            }\n        }\n        \"#,\nSome(CompletionKind::Reference))"
---
lc self Foo
sp Self
st Foo
